### DISCUSSION TITLE
OSSL_LIB_CTX_load_config loads system config even when a particular config is specified

### DISCUSSION THREAD

**Discussion by beldmit** [openssl/openssl#23720] [Q&A] [answered]
This is a follow-up to https://github.com/open-quantum-safe/oqs-provider/issues/358 and especially https://github.com/open-quantum-safe/oqs-provider/issues/358#issuecomment-1973049693

I came across that oqs provider tests read system config despite having the local one tuned for test purposes, and it affects tests. It is easy to fix the tests but I wonder if the behavior is intentional.

The stack trace leading to loading system config is

```
bt
#0  _IO_new_fopen (filename=filename@entry=0x419170 "/etc/pki/tls/openssl.cnf", mode=mode@entry=0x7ffff7d8c1c9 "rb") at iofopen.c:85
#1  0x00007ffff7a7e470 in openssl_fopen (mode=0x7ffff7d8c1c9 "rb", filename=0x419170 "/etc/pki/tls/openssl.cnf") at crypto/o_fopen.c:112
#2  BIO_new_file (filename=0x419170 "/etc/pki/tls/openssl.cnf", mode=0x7ffff7d8c1c9 "rb") at crypto/bio/bss_file.c:60
#3  0x00007ffff7abb3f6 in def_load (conf=0x418dc0, name=<optimized out>, line=0x0) at crypto/conf/conf_def.c:175
#4  0x00007ffff7abc3a2 in CONF_modules_load_file_ex (libctx=<optimized out>, filename=0x0, appname=0x0, flags=50) at crypto/conf/conf_mod.c:206
#5  0x00007ffff7b71fc8 in ossl_config_int (settings=0x0) at crypto/conf/conf_sap.c:68
#6  ossl_config_int (settings=0x0) at crypto/conf/conf_sap.c:44
#7  ossl_init_config () at crypto/init.c:256
#8  ossl_init_config_ossl_ () at crypto/init.c:254
#9  0x00007ffff78ae50b in __pthread_once_slow (once_control=0x7ffff7ea950c <config>, init_routine=0x7ffff7b71fa0 <ossl_init_config_ossl_>) at pthread_once.c:116
#10 0x00007ffff78ae579 in ___pthread_once (once_control=<optimized out>, init_routine=<optimized out>) at pthread_once.c:143
#11 0x00007ffff7b7e9dd in CRYPTO_THREAD_run_once (once=<optimized out>, init=<optimized out>) at crypto/threads_pthread.c:152
#12 0x00007ffff7b72a9b in OPENSSL_init_crypto (opts=64, settings=0x0) at crypto/init.c:599
#13 0x00007ffff7b8ba4c in ossl_init_added_lock () at crypto/objects/obj_dat.c:78
#14 ossl_obj_read_lock (lock=1) at crypto/objects/obj_dat.c:96
#15 OBJ_sn2nid (s=s@entry=0x7ffff77c9000 "1.3.9999.99.13") at crypto/objects/obj_dat.c:663
#16 0x00007ffff7b8bb41 in OBJ_txt2obj (s=0x7ffff77c9000 "1.3.9999.99.13", no_name=no_name@entry=0) at crypto/objects/obj_dat.c:421
#17 0x00007ffff7b8bcb5 in OBJ_txt2nid (s=<optimized out>) at crypto/objects/obj_dat.c:614
#18 0x00007ffff7b82132 in core_obj_create (prov=<optimized out>, oid=0x7ffff77c9000 "1.3.9999.99.13", sn=0x7ffff77c900f "frodo640aes", ln=0x7ffff77c900f "frodo640aes") at crypto/provider_core.c:2225
#19 0x00007ffff777fed0 in OSSL_provider_init (handle=0x418af0, in=0x7ffff7e9f630 <core_dispatch_+816>, out=0x7fffffffd988, provctx=0x7fffffffd990) at /root/oqs-provider/oqsprov/oqsprov.c:968
#20 0x00007ffff7b81692 in provider_init (prov=0x418af0) at crypto/provider_core.c:964
#21 provider_activate (prov=prov@entry=0x418af0, lock=0, lock@entry=1, upcalls=upcalls@entry=1) at crypto/provider_core.c:1173
#22 0x00007ffff7b81bbf in ossl_provider_activate (prov=prov@entry=0x418af0, upcalls=upcalls@entry=1, aschild=aschild@entry=0) at crypto/provider_core.c:1310
#23 0x00007ffff7b82eff in provider_conf_activate (libctx=libctx@entry=0x4072a0, name=name@entry=0x419910 "oqsprovider", value=value@entry=0x419960 "oqsprovider_sect", path=path@entry=0x0, soft=soft@entry=0, cnf=cnf@entry=0x416a00)
    at crypto/provider_conf.c:245
#24 0x00007ffff7b83281 in provider_conf_load (libctx=0x4072a0, name=<optimized out>, name@entry=0x419910 "oqsprovider", value=value@entry=0x419960 "oqsprovider_sect", cnf=cnf@entry=0x416a00) at crypto/provider_conf.c:323
#25 0x00007ffff7b8349b in provider_conf_init (md=<optimized out>, cnf=0x416a00) at crypto/provider_conf.c:379
#26 0x00007ffff7abbe15 in module_init (cnf=<optimized out>, value=0x419770 "provider_sect", name=0x419720 "providers", pmod=0x418420) at crypto/conf/conf_mod.c:425
#27 module_run (flags=<optimized out>, value=<optimized out>, name=0x419720 "providers", cnf=<optimized out>) at crypto/conf/conf_mod.c:272
#28 CONF_modules_load (cnf=cnf@entry=0x416a00, appname=appname@entry=0x0, flags=flags@entry=0) at crypto/conf/conf_mod.c:165
#29 0x00007ffff7abc3b8 in CONF_modules_load_file_ex (libctx=<optimized out>, filename=0x7fffffffe12b "/root/oqs-provider/build/../test/openssl-ca.cnf", appname=0x0, flags=0) at crypto/conf/conf_mod.c:214
#30 0x00007ffff7b6e531 in OSSL_LIB_CTX_load_config (ctx=<optimized out>, config_file=<optimized out>) at crypto/context.c:515
#31 0x0000000000402c42 in load_oqs_provider (libctx=0x4072a0, modulename=0x7fffffffe11f "oqsprovider", configfile=0x7fffffffe12b "/root/oqs-provider/build/../test/openssl-ca.cnf") at /root/oqs-provider/test/test_common.c:50
#32 0x00000000004029f4 in main (argc=4, argv=0x7fffffffdda8) at /root/oqs-provider/test/oqs_test_tlssig.c:166
```

**Comment by mattcaswell** [comment]
The `OBJ_*` code is not libctx aware. You can add objects to the OBJ database from config and they always get initialised from the system config file regardless. In this case the libctx specific config file is loading the OQS provider which ends up calling an `OBJ_*` function which then checks the system config file to see if any objects should be loaded from config.

**Reply by beldmit** [reply]
The problem I came across is that Red Hat/Fedora has limitations in our system-level config that interferes with the test. I understand that we can specify `OPENSSL_CONF` variable in the tests, but I wonder if we should consider it as a workaround or as a legitimate solution?

**Reply by mattcaswell** [reply]
The long term answer is to make everything needing the config file be libctx aware. This is rather a missing feature than a bug though.

Does it matter if the system file config fails to load? It should just mean that you don't get any OBJ objects loaded from config - which in most cases isn't a problem.

**Reply by beldmit** [reply]
The problem is that it **succeeded** to load - and so tests failed (SignatureAlgorithms from system config were applied). At least it was unexpected.

**Reply by mattcaswell** [reply]
Hmm...but surely it only impacts tests which are using the default libctx and would have loaded the system config file anyway?

**Reply by beldmit** [reply]
https://github.com/open-quantum-safe/oqs-provider/blob/c442a5c30548a0c2740e106dafbf44b678e6c397/test/test_common.c#L47-L52 is the code. It looks like a viable pattern. And yes, attempting initializing the provider from the config causes system-wide config loading

**Reply by mattcaswell** [reply]
Yes - i get that (for the reasons I described above). My question is why does this cause a problem? Except for the OBJ database, loading the system-wide config should only have an impact on anything that susbsequently uses the default libctx - and those things would have eventually loaded the system wide config file anyway. It just means that it is happening earlier than expected in a slightly surprising way.

**Reply by beldmit** [reply]
Sorry, I didn't get your question initially. 

The test config doesn't apply any limits to TLS configuration. The system config applies crypto-policies and limits SignatureAlgorithms to well-known algorithms overriding (absence of) this limitation in the test config

**Reply by t8m** [reply]
You have to explicitly call `OPENSSL_init_crypto/ssl(OPENSSL_INIT_NO_LOAD_CONFIG,NULL);` to prevent the global default libctx being initialized from the default config file.

**Reply by beldmit** [reply]
Is the current behavior desired or not?

**Reply by t8m** [reply, marked as answer]
I do not think we can change it. At least not before 4.0.

**Comment by nhorman** [comment]
As a work around you could just run the tests in a container with the system config bind mounted to an empty file (or other appropriate file)

**Reply by beldmit** [reply]
Sure, just explicit export OPENSSL_CONF=/dev/null works
